src/rust/engine/src/scheduler.rs

```diff
@@ -261,12 +267,54 @@ impl Scheduler {
       core: self.core.clone(),
     };
     let (sender, receiver) = mpsc::channel();
+
+    // Setting up display
+    let display_worker_count = 8;
+    let mut optional_display = if request.should_render_ui {
+      Some(EngineDisplay::for_stdout(0))
+    } else {
+      None
+    };
+    if let Some(display) = optional_display.as_mut() {
+      display.start();
+      display.render();
+      let worker_ids: Vec<String> = (0..display_worker_count)
+        .map(|s| format!("{}", s))
+        .collect();
+      for worker_id in worker_ids {
+        display.add_worker(worker_id);
+        display.render();
+      }
+    }
+
     Scheduler::execute_helper(context, sender, request.roots.clone(), 8);
+    let roots: Vec<_> = request
+      .roots
+      .clone()
+      .into_iter()
+      .map(|s| s.into())
+      .collect();
+
     let results = loop {
       if let Ok(res) = receiver.recv_timeout(Duration::from_millis(100)) {
         break res;
+      } else if let Some(display) = optional_display.as_mut() {
+        let ongoing_tasks = self.core.graph.heavy_hitters(&roots, display_worker_count);
+        for (i, task) in ongoing_tasks.iter().enumerate() {
+          display.update(i.to_string(), format!("{:?}", task));
+        }
+        // If the number of ongoing tasks is less than the number of workers,
+        // fill the rest of the workers with empty string.
```

Comment:
Can you open a new ticket about doing something position-preserving (ie, the 3 options described in the ticket) and reference it here?

Reply:
linked.